specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,471 @@
+and to prevent footguns when trying to deposit an ERC20 native asset into a chain that uses `ether` as its native
+asset.
+
+```solidity
+function depositERC20Transaction(
+    address _to,
+    uint256 _mint,
+    uint256 _value,
+    uint64 _gasLimit,
+    bool _isCreation,
+    bytes memory _data
+) public;
+```
+
+This function MUST revert when `ether` is the L2's native asset. It MUST not be `payable`, meaning that it will
+revert when `ether` is sent with a `CALL` to it. It uses a `transferFrom` flow, so users MUST first `approve()`
+the `OptimismPortal` before they can deposit tokens.
+
+##### Function Arguments
+
+The following table describes the arguments to `depositERC20Transaction`
+
+| Name | Type | Description |
+|---------|---------|---------|
+| `_to` | `address` | The target of the deposit transaction |
+| `_mint` | `uint256` | The amount of token to deposit |
+| `_value` | `uint256` | The value of the deposit transaction, used to transfer native asset that is already on L2 from L1 |
+| `_gasLimit` | `uint64` | The gas limit of the deposit transaction |
+| `_isCreation` | `bool` | Signifies the `_data` should be used with `CREATE` |
+| `_data` | `bytes` | The calldata of the deposit transaction |
+
+#### `depositTransaction`
+
+This function is the hot code path for all usage of the `L1StandardBridge` and `L1CrossDomainMessenger`, so it MUST
+be as backwards compatible as possible. It MUST revert for chains using custom gas token when `ether` is sent with
+the `CALL`. This prevents `ether` from being stuck in the `OptimismPortal`, since it cannot be used to mint native
+asset on L2.
+
+#### `setGasPayingToken`
+
+This function MUST only be callable by the `SystemConfig`. When called, it creates a special deposit transaction
+from the `DEPOSITOR_ACCOUNT` that calls the `L1Block.setGasPayingToken` function. The ERC20 `name` and `symbol`
+are passed as `bytes32` to prevent the usage of dynamically sized `string` arguments.
+
+```solidity
+function setGasPayingToken(address _token, uint8 _decimals, bytes32 _name, bytes32 _symbol) external;
+```
+
+### StandardBridge
+
+The `StandardBridge` contracts on both L1 and L2 MUST be aware of the custom gas token. The `ether` specific ABI on the
+`StandardBridge` is disabled when custom gas token is enabled.
+
+The following methods MUST revert when custom gas token is being used:
+
+- `bridgeETH(uint32,bytes)`
+- `bridgeETHTo(address,uint32,bytes)`
+- `receive()`
+- `finalizeBridgeETH(address, address, uint256, bytes)`
+
+The following legacy methods MUST also revert when custom gas token is being used:
+
+- `depositETH(uint32,bytes)`
+- `depositETHTo(address,uint32,bytes)`
+- `finalizeETHWithdrawal(address,address,uint256,bytes)`
+- `withdraw(address,uint256,uint32,bytes)`
+- `withdrawTo(address,address,uint256,uint32,bytes)`
+- `finalizeDeposit(address,address,address,address,uint256,bytes)`
```

Comment:
Why do these ones need to revert? They don't necessarily involve transfer of the gas token right?

Also nit, these 3 methods are only in L2StandardBridge, and 3 above are only in the L1StandardBridge—can we clarify that a bit? We're under a `StandardBridge` header so this is slightly misleading

Reply:
I believe it should say they revert if non zero callvalue

Reply:
originally we had it that the revert would happen for non-zero callvalue, but then we decided to disable the methods that have `ETH` in their names altogether when a custom gas token is used

added the headers for `L1StandardBridge`, `L2StandardBridge`

Reply:
Here is the updated part of the spec, annotated with my thoughts/open questions:

---

The following methods MUST revert when custom gas token is being used:

- `bridgeETH(uint32,bytes)`: ✅ Agreed, because this deals with ETH
- `bridgeETHTo(address,uint32,bytes)`: ✅ Agreed, because this deals with ETH
- `receive()`: ✅ Agreed, because this deals with ETH
- `finalizeBridgeETH(address, address, uint256, bytes)`: ✅ Agreed, because this deals with ETH

The following legacy methods MUST also revert when custom gas token is being used:

- In `L1StandardBridge`:
    - `depositETH(uint32,bytes)`: ✅ Agreed, because this deals with ETH
    - `depositETHTo(address,uint32,bytes)`: ✅ Agreed, because this deals with ETH
    - `finalizeETHWithdrawal(address,address,uint256,bytes)`: ✅ Agreed, because this deals with ETH
- In `L2StandardBridge`:
    - `withdraw(address,uint256,uint32,bytes)`: ❓ Unclear on rationale for this one
    - `withdrawTo(address,address,uint256,uint32,bytes)`: ❓ Unclear on rationale for this one
    - `finalizeDeposit(address,address,address,address,uint256,bytes)`: ❓ Unclear on rationale for this one

Reply:
The unclear ones are because they can be used for sending ether between domains. They also work with other tokens which is why it should be clear that they revert when `CALLVALUE` is non zero

Reply:
I see, so I think it'd be more precise for the spec to read like the below? (maybe a nit here, just want to make sure the spec is clear)

---

The following legacy methods in `L1StandardBridge` MUST revert when custom gas token is being used:

- `depositETH(uint32,bytes)`
- `depositETHTo(address,uint32,bytes)`
- `finalizeETHWithdrawal(address,address,uint256,bytes)`

The following legacy methods in `L2StandardBridge` MUST revert when custom gas token is being used and the `CALLVALUE` is nonzer:
- `withdraw(address,uint256,uint32,bytes)`
- `withdrawTo(address,address,uint256,uint32,bytes)`
- `finalizeDeposit(address,address,address,address,uint256,bytes)`

Reply:
good idea!

Reply:
This has been resolved